File: mcdowell/python/ch18-08.py


class PrefixTree:
    root = None

    def __init__(self, string):
        self.root = PrefixTreeNode()
        for i in range(len(string)):
            self.root.insertString(string[i:], i)


    def search(self, string):
        return self.root.search(string)


class PrefixTreeNode:
    children = None
    indexes = None

    def __init__(self):
        self.children = {}
        self.indexes = []


    def insertString(self, string, index):
        if string is not None and len(string) > 0:
            firstLetter = string[0]
            child = None

            if firstLetter in self.children:
                child = self.children[firstLetter]
            else:
                child = PrefixTreeNode()
                self.children[firstLetter] = child

            child.indexes.append(index)
            child.insertString(string[1:], index)



    def search(self, string):
        if string is None or len(string) == 0:
            return self.indexes

        firstLetter = string[0]
        if firstLetter in self.children:
            return self.children[firstLetter].search(string[1:])

        return -1

    def preOrderTranverse(self, depth = 0):

        padStr = ""
        for i in range(depth):
            padStr += "-"

        for k in self.children:
            print padStr + k
            self.children[k].preOrderTranverse(depth + 2)



tree = PrefixTree("bibs")

tree.root.preOrderTranverse()


print tree.search("s")

File: mcdowell/php/ch18-08.php
<?php
// prefix tree: search for substring in one string
// suffix tree: search for longest common substring between many strings

class PrefixTree {
    public $root = null;
    public function __construct($string) {
        $this->root = new PrefixTreeNode();
        for ($i = 0; $i < strlen($string); $i ++) {
            $this->root->insertString(substr($string, $i), $i);
        }
    }

    public function search($string) {
        return $this->root->search($string);
    }
}


class PrefixTreeNode {
    public $children = []; // edges store letter
    public $indexes = []; // vertices store starting index

    public function insertString($string, $index) {
        if ($string != null && strlen($string)) {
            $value = $string{0};
            $child = null;
            if (isset($this->children[$value])) {
                $child = $this->children[$value];
            }
            else {
                $child = new PrefixTreeNode();
                $this->children[$value] = $child;
            }

            $remainder = substr($string, 1);
            $child->indexes[] = $index;
            $child->insertString($remainder, $index);
        }
    }

    public function search($string) {
        if ($string == null || !strlen($string)) {
            return $this->indexes;
        }
        else {
            $first = $string{0};
            if (isset($this->children[$first])) {
                $reminder = substr($string, 1);
                return $this->children[$first]->search($reminder);
            }

        }
        return null;
    }

    public function preOrderTranverse($depth = 0) {

        $pad = str_pad("", $depth, '-');


        foreach($this->children AS $k => $child) {
            echo "$pad{$k}     ";
            echo "index: " . implode(', ', $child->indexes);
            echo "\n";
            $child->preOrderTranverse($depth + 2);
        }


    }
}



$tree = new PrefixTree("bibs");
$tree->root->preOrderTranverse();

print_r($tree->search('bibs'));